lib/spack/spack/installer.py

```diff
@@ -0,0 +1,1554 @@
+    """
+    Colorize the name/id of the package being installed
+
+    Args:
+        name (str): Name/id of the package being installed
+
+    Return:
+        (str) Colorized installing message
+    """
+    return colorize('@*{Installing} @*g{%s}' % name)
+
+
+def log(pkg):
+    """
+    Copy provenance into the install directory on success
+
+    Args:
+        pkg (Package): the package to be installed and built
+    """
+    packages_dir = spack.store.layout.build_packages_path(pkg.spec)
+
+    # Remove first if we're overwriting another build
+    # (can happen with spack setup)
+    try:
+        # log and env install paths are inside this
+        shutil.rmtree(packages_dir)
+    except Exception as e:
+        # FIXME : this potentially catches too many things...
+        tty.debug(e)
+
+    # Archive the whole stdout + stderr for the package
+    install(pkg.log_path, pkg.install_log_path)
+
+    # Archive the environment used for the build
+    install(pkg.env_path, pkg.install_env_path)
+
+    # Finally, archive files that are specific to each package
+    with working_dir(pkg.stage.path):
+        errors = StringIO()
+        target_dir = os.path.join(
+            spack.store.layout.metadata_path(pkg.spec), 'archived-files')
+
+        for glob_expr in pkg.archive_files:
+            # Check that we are trying to copy things that are
+            # in the stage tree (not arbitrary files)
+            abs_expr = os.path.realpath(glob_expr)
+            if os.path.realpath(pkg.stage.path) not in abs_expr:
+                errors.write('[OUTSIDE SOURCE PATH]: {0}\n'.format(glob_expr))
+                continue
+            # Now that we are sure that the path is within the correct
+            # folder, make it relative and check for matches
+            if os.path.isabs(glob_expr):
+                glob_expr = os.path.relpath(glob_expr, pkg.stage.path)
+            files = glob.glob(glob_expr)
+            for f in files:
+                try:
+                    target = os.path.join(target_dir, f)
+                    # We must ensure that the directory exists before
+                    # copying a file in
+                    mkdirp(os.path.dirname(target))
+                    install(f, target)
+                except Exception as e:
+                    tty.debug(e)
+
+                    # Here try to be conservative, and avoid discarding
+                    # the whole install procedure because of copying a
+                    # single file failed
+                    errors.write('[FAILED TO ARCHIVE]: {0}'.format(f))
+
+        if errors.getvalue():
+            error_file = os.path.join(target_dir, 'errors.txt')
+            mkdirp(target_dir)
+            with open(error_file, 'w') as err:
+                err.write(errors.getvalue())
+            tty.warn('Errors occurred when archiving files.\n\t'
+                     'See: {0}'.format(error_file))
+
+    dump_packages(pkg.spec, packages_dir)
```

Comment:
https://github.com/spack/spack/blob/ca26eb69231fffb500459746a218b04724657652/lib/spack/spack/package.py#L1904-L1969

Reply:
Similar to `dump_packages`?